### PR TITLE
client: Add crio client package

### PR DESCRIPTION
**Pull request by mrunalp**
This PR adds a client package for accessing the crio info and container endpoints. We also expose all the crio annotations in the container endpoint.
Signed-off-by: Mrunal Patel <mrunalp@gmail.com>

### CODE REVIEW

**Comment by mrunalp** [comment]
@bparees This is what you get in the container endpoint. You need to get the ResolvPath from crio_annotations.
```
crioclient container 494b682ff63e3a3fa7702537e8a84ae92b5e945c2276f93cf061a0c4d3095985
{
    "name": "k8s_kubedns_kube-dns-3097350089-49d91_kube-system_5c8c71de-9fef-11e7-8181-54ee75105a2b_0",
    "pid": 23629,
    "image": "407bc2fe5e28aadeb7da42758ba838df8573fd24ec3c4397d2b93a2dc86b125f",
    "created_time": 1506123607353159411,
    "labels": {
        "io.kubernetes.container.name": "kubedns",
        "io.kubernetes.pod.name": "kube-dns-3097350089-49d91",
        "io.kubernetes.pod.namespace": "kube-system",
        "io.kubernetes.pod.uid": "5c8c71de-9fef-11e7-8181-54ee75105a2b"
    },
    "annotations": {
        "io.kubernetes.container.hash": "abff0251",
        "io.kubernetes.container.ports": "[{\"name\":\"dns-local\",\"containerPort\":10053,\"protocol\":\"UDP\"},{\"name\":\"dns-tcp-local\",\"containerPort\":10053,\"protocol\":\"TCP\"},{\"name\":\"metrics\",\"containerPort\":10055,\"protocol\":\"TCP\"}]",
        "io.kubernetes.container.restartCount": "0",
        "io.kubernetes.container.terminationMessagePath": "/dev/termination-log",
        "io.kubernetes.container.terminationMessagePolicy": "File",
        "io.kubernetes.pod.terminationGracePeriod": "30"
    },
    "crio_annotations": {
        "io.kubernetes.container.hash": "abff0251",
        "io.kubernetes.container.name": "kubedns",
        "io.kubernetes.container.ports": "[{\"name\":\"dns-local\",\"containerPort\":10053,\"protocol\":\"UDP\"},{\"name\":\"dns-tcp-local\",\"containerPort\":10053,\"protocol\":\"TCP\"},{\"name\":\"metrics\",\"containerPort\":10055,\"protocol\":\"TCP\"}]",
        "io.kubernetes.container.restartCount": "0",
        "io.kubernetes.container.terminationMessagePath": "/dev/termination-log",
        "io.kubernetes.container.terminationMessagePolicy": "File",
        "io.kubernetes.cri-o.Annotations": "{\"io.kubernetes.container.hash\":\"abff0251\",\"io.kubernetes.container.ports\":\"[{\\\"name\\\":\\\"dns-local\\\",\\\"containerPort\\\":10053,\\\"protocol\\\":\\\"UDP\\\"},{\\\"name\\\":\\\"dns-tcp-local\\\",\\\"containerPort\\\":10053,\\\"protocol\\\":\\\"TCP\\\"},{\\\"name\\\":\\\"metrics\\\",\\\"containerPort\\\":10055,\\\"protocol\\\":\\\"TCP\\\"}]\",\"io.kubernetes.container.restartCount\":\"0\",\"io.kubernetes.container.terminationMessagePath\":\"/dev/termination-log\",\"io.kubernetes.container.terminationMessagePolicy\":\"File\",\"io.kubernetes.pod.terminationGracePeriod\":\"30\"}",
        "io.kubernetes.cri-o.ContainerID": "494b682ff63e3a3fa7702537e8a84ae92b5e945c2276f93cf061a0c4d3095985",
        "io.kubernetes.cri-o.ContainerType": "container",
        "io.kubernetes.cri-o.Created": "2017-09-22T16:40:07.051243146-07:00",
        "io.kubernetes.cri-o.IP": "10.88.22.10",
        "io.kubernetes.cri-o.Image": "407bc2fe5e28aadeb7da42758ba838df8573fd24ec3c4397d2b93a2dc86b125f",
        "io.kubernetes.cri-o.ImageName": "gcr.io/google_containers/k8s-dns-kube-dns-amd64:1.14.4",
        "io.kubernetes.cri-o.ImageRef": "407bc2fe5e28aadeb7da42758ba838df8573fd24ec3c4397d2b93a2dc86b125f",
        "io.kubernetes.cri-o.Labels": "{\"io.kubernetes.container.name\":\"kubedns\",\"io.kubernetes.pod.name\":\"kube-dns-3097350089-49d91\",\"io.kubernetes.pod.namespace\":\"kube-system\",\"io.kubernetes.pod.uid\":\"5c8c71de-9fef-11e7-8181-54ee75105a2b\"}",
        "io.kubernetes.cri-o.LogPath": "/var/log/pods/5c8c71de-9fef-11e7-8181-54ee75105a2b/kubedns_0.log",
        "io.kubernetes.cri-o.Metadata": "{\"name\":\"kubedns\"}",
        "io.kubernetes.cri-o.MountPoint": "/var/lib/containers/storage/overlay/e33f8f0eb9878bad6301a7a4439601d3d8ef9f89b53eea33dd827f7a78f37039/merged",
        "io.kubernetes.cri-o.Name": "k8s_kubedns_kube-dns-3097350089-49d91_kube-system_5c8c71de-9fef-11e7-8181-54ee75105a2b_0",
        "io.kubernetes.cri-o.ResolvPath": "/var/run/containers/storage/overlay-containers/e5a667df3e363da102e7777c449b0dcc5d37058a2d74ff71110b073bd135b72f/userdata/resolv.conf",
        "io.kubernetes.cri-o.SandboxID": "e5a667df3e363da102e7777c449b0dcc5d37058a2d74ff71110b073bd135b72f",
        "io.kubernetes.cri-o.SandboxName": "k8s_POD_kube-dns-3097350089-49d91_kube-system_5c8c71de-9fef-11e7-8181-54ee75105a2b_0",
        "io.kubernetes.cri-o.Stdin": "false",
        "io.kubernetes.cri-o.StdinOnce": "false",
        "io.kubernetes.cri-o.TTY": "false",
        "io.kubernetes.cri-o.Volumes": "[{\"container_path\":\"/kube-dns-config\",\"host_path\":\"/var/lib/kubelet/pods/5c8c71de-9fef-11e7-8181-54ee75105a2b/volumes/kubernetes.io~configmap/kube-dns-config\",\"readonly\":false},{\"container_path\":\"/var/run/secrets/kubernetes.io/serviceaccount\",\"host_path\":\"/var/lib/kubelet/pods/5c8c71de-9fef-11e7-8181-54ee75105a2b/volumes/kubernetes.io~secret/kube-dns-token-hf3ck\",\"readonly\":true},{\"container_path\":\"/etc/hosts\",\"host_path\":\"/var/lib/kubelet/pods/5c8c71de-9fef-11e7-8181-54ee75105a2b/etc-hosts\",\"readonly\":false},{\"container_path\":\"/dev/termination-log\",\"host_path\":\"/var/lib/kubelet/pods/5c8c71de-9fef-11e7-8181-54ee75105a2b/containers/kubedns/9df8e98c\",\"readonly\":false}]",
        "io.kubernetes.pod.name": "kube-dns-3097350089-49d91",
        "io.kubernetes.pod.namespace": "kube-system",
        "io.kubernetes.pod.terminationGracePeriod": "30",
        "io.kubernetes.pod.uid": "5c8c71de-9fef-11e7-8181-54ee75105a2b"
    },
    "log_path": "/var/log/pods/5c8c71de-9fef-11e7-8181-54ee75105a2b/kubedns_0.log",
    "root": "/var/lib/containers/storage/overlay/e33f8f0eb9878bad6301a7a4439601d3d8ef9f89b53eea33dd827f7a78f37039/merged",
    "sandbox": "e5a667df3e363da102e7777c449b0dcc5d37058a2d74ff71110b073bd135b72f",
    "ip_address": "10.88.22.10"
}
```

**Comment by runcom** [comment]
Looks good

**Comment by rhatdan** [comment]
What is the purpose of this tool?  Is this something we expect users to use?  Or is it more for testing?  Why wasn't the functionality added to crioctl?  If it is a new tool for users we need to add at least a man page for it.  If it is not intended for use by admin, then we should not install it or not put in in /usr/bin.

**Comment by runcom** [comment]
I believe we won't install it, the golang pkg is needed by openshift instead

**Comment by rhatdan** [comment]
Well if it is only going to be used by another service like openshift, then it should be installed in /usr/libexec not in /usr/bin.  Then we don't need a man page for it.

**Comment by runcom** [comment]
Sounds good, but I'd rather not installing it at all

**Comment by mrunalp** [comment]
I think we can just fold this into crioctl. I will update the PR to do that.

**Comment by mrunalp** [comment]
Removed crioclient and added the commands to crioctl. I will add a test next.

**Comment by bparees** [comment]
@mrunalp i'm not gettin the ResolvPath annotation in my ClientInfo:

```
crio.ContainerInfo{Name:"k8s_sti-build_nodejs-ex-4-build_p1_dcbe3c4f-a087-11e7-a2d0-507b9d27b5d9_0", Pid:29698, Image:"1b4b847a5a7cac552955c0b95af954bbe64ee1be671dd32b244fa65d10c21251", CreatedTime:1506189123852724479, Labels:map[string]string{"io.kubernetes.container.name":"sti-build", "io.kubernetes.pod.name":"nodejs-ex-4-build", "io.kubernetes.pod.namespace":"p1", "io.kubernetes.pod.uid":"dcbe3c4f-a087-11e7-a2d0-507b9d27b5d9"}, Annotations:map[string]string{"io.kubernetes.container.terminationMessagePath":"/dev/termination-log", "io.kubernetes.container.terminationMessagePolicy":"FallbackToLogsOnError", "io.kubernetes.pod.terminationGracePeriod":"30", "io.kubernetes.container.hash":"98ef69d7", "io.kubernetes.container.restartCount":"0"}, LogPath:"/var/log/pods/dcbe3c4f-a087-11e7-a2d0-507b9d27b5d9/sti-build_0.log", Root:"/var/lib/containers/storage/overlay/abe7543c99c4694583158d7695b6a4b72ce545f31f4052e28b254310092fc887/merged", IP:"10.88.0.117"}
```

**Comment by mrunalp** [comment]
@bparees It should be in the `CrioAnnotations` variable. Yesterday I realized that we were only returning kube specific annotations so added this new variable to the struct in this PR.
https://github.com/kubernetes-incubator/cri-o/pull/943/files#diff-9457129899e29a9925770fb1644a14f3R11

**Comment by mrunalp** [comment]
Added steps to the inspect test to use the client as well.

File: types/types.go [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+// ContainerInfo stores information about containers
+type ContainerInfo struct {
+	Name            string            `json:"name"`
+	Pid             int               `json:"pid"`
+	Image           string            `json:"image"`
+	CreatedTime     int64             `json:"created_time"`
+	Labels          map[string]string `json:"labels"`
+	Annotations     map[string]string `json:"annotations"`
+	CrioAnnotations map[string]string `json:"crio_annotations"`
```
Comment (bparees) [inline types/types.go]: I added this field in my copy of this struct but i'm not seeing what is supposed to populate the value in the server?

https://github.com/kubernetes-incubator/cri-o/blob/master/server/inspect.go#L68-L79

Is there another patch to cri-o that i'm missing?

Comment (bparees) [reply types/types.go]: also note that origin already vendors a version of this client code, as part of vendoring cadvisor:

https://github.com/openshift/origin/blob/master/vendor/github.com/google/cadvisor/container/crio/client.go#L38-L48

So should cadvisor's containerinfo struct also be patched with this new annotation?  should this client code not go into cri-o and it should stay in cadvisor?

**Comment by mrunalp** [comment]
The second commit in this PR adds crio annotation support. We need this entire PR for the client and resolvpath annotation.

**Comment by mrunalp** [comment]
We will change cadvisor to use this new client in this PR but for now it can work as is as it doesn't  need any crio specific annotations.

**Comment by mrunalp** [comment]
@cevich This is ready for a manual test run.

**Comment by rhatdan** [comment]
We eventually want to drop crioctl for crictl, correct?  And do we want crioctl for anything more then a test environment and other tools like openshift to execute?  If so we probably want to move it out of /usr/bin/also.

**Comment by mrunalp** [comment]
Yes, even openshift won't be executing it directly. They are going to use the client pkg. We are using it to test the client package.

**Comment by rhatdan** [comment]
So we should not install the executable at all.

**Comment by mrunalp** [comment]
Well, we are suggesting it as the primary tool to debug cri-o till kpod is ready. We could take it out in a future release.

**Comment by mrunalp** [comment]
@runcom The tests passed here. Can we get this merged?

**Comment by smarterclayton** [comment]
Is there a reason this wasn't added to the existing container status endpoint?  As is this isn't something that is accessible via CRI, so we have to have a custom client and library that has to be accessed differently than CRI

**Comment by runcom** [comment]
The only user of this is cadvisor which is going to be deprecated in favor of stats from CRI. We thought not to clutter the CRI and go this was till cadvisor is gkne

**Comment by runcom** [comment]
The only user of this is cadvisor which is going to be deprecated in favor of stats from CRI. We thought not to clutter the CRI and go this was till cadvisor is gone

**Pull request merged**